### ISSUE TITLE
How should the client act when textDocumentSync=none?

### ISSUE THREAD

**Issue by LevoD** [Microsoft/language-server-protocol#722]
Hello everyone.

I'm implementing a proof of concept language server. I was playing around with textDocument/hover. On a document I didn't edit I hovered over a specific word and then hit return so it'd be on a new line. I hovered over it again and checked what VS code sent me. The word was on a new line even though on "initialize" textDocumentSync is set to none (actually it's not set at all but it's not behaving as it did when I had it set to full and incremental, so I'm assuming this is none behavior).

If the language server isn't getting any text document sync's shouldn't the client/vscode either A) track where the word is in the original document and send that location? or if it doesn't want to do that B) Not send the hover request at all? Is it documented what should happen when the server doesn't handle document syncs? I may have missed a warning about not supporting text document syncs.

I suspect it wouldn't be too hard to translate current text position to original but I'm not sure if the client/editor should do it or if the server should. Either way I thought this behavior was strange and would be surprising to a user

**Comment by Aerijo** [comment]
>  hit return so it'd be on a new line. I hovered over it again and checked what VS code sent me. The word was on a new line

Yes. I'd expect this every time, regardless of if the server wants to sync with the client. A better question is why would the client ever try to remember the original state of the document?

> or if it doesn't want to do that B) Not send the hover request at all?

If your server doesn't support hover requests then the client won't request them.

> Is it documented what should happen when the server doesn't handle document syncs?

I don't think this needs to be documented. From the clients point of view, they're doing all they can to let the server know the current state of the document. If the server can't handle that, then the server needs to find a solution the problem, not the client or the spec. And that solution is most likely to be supporting sync notifications. Because that's the only way to know the current unsaved state of the file.

> I suspect it wouldn't be too hard to translate current text position to original

I suspect it would be impossible. You are talking about tracking arbitrary changes, including replacing the contents with something entirely unrelated yes?

> Either way I thought this behavior was strange and would be surprising to a user

I think this theoretical server's behaviour is strange, and would be surprising to the user AND the client.

**Comment by dbaeumer** [comment]
@Aerijo thanks for the great explanation. 

Some additions: the `textDocumentSync=none` makes sense in read only cases since no content every travels over to the server (remember that an open notification sends the whole file content). The server can handle everything on disk.

I will close the issue since as @Aerijo pointed out as well this is nothing the spec should address. 

@LevoD please ping if you strongly disagree.

**Comment by LevoD** [comment]
> @LevoD please ping if you strongly disagree.

I completely and strongly disagree with Aerijo.

> A better question is why would the client ever try to remember the original state of the document?

VS Code does it as an optimization https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation but that's besides the point. I never said it should remember the original file. I said it should either track where the row:col would be in the original file or not send request on dirty documents.

> If your server doesn't support hover requests then the client won't request them.

It does. Which is my point

(I'll jump ahead a paragraph to address this first)

> > I suspect it wouldn't be too hard to translate current text position to original

> I suspect it would be impossible

From the server prescriptive absolutely. Would be trivial for the client to mark a document as dirty and not send events anymore.

> I don't think this needs to be documented. From the clients point of view, they're doing all they can to let the server know the current state of the document. If the server can't handle that, then the server needs to find a solution the problem, not the client or the spec. And that solution is most likely to be supporting sync notifications. Because that's the only way to know the current unsaved state of the file.

I don't think it's in the spirit of the textDocumentSync capability to enable it just to mark the file as dirty and to ignore future request to that document. I haven't test what other clients do but it'd be completely reasonable for them to either send the original lines or to stop sending request to that document. Which client would be right? As as Aerijo said it's impossible for a server to know if a document is dirty when textDocumentSync is set to none. 

If the reason TextDocumentSyncKind.None exist is to send fewer request then it's inconsistent to send request to hover and anything else that depends on lines when a document is dirty.

> I think this theoretical server's behaviour is strange, and would be surprising to the user AND the client.

Don't act like it's not surprising to send request to handle a situation you yourself called impossible.

To sum this up
- If TextDocumentSyncKind.None exist to send fewer request it should be more consistent and not send request that are impossible to handle  
- It'd be reasonable for clients to not send request to dirty documents or to do better and send the original line:columns. Right now it's unspecified so two clients may do it differently  
- I don't think claiming to handle document syncing just to mark a document as dirty is a good idea
- TextDocumentSyncKind.None might be completely useless in this current state. But I haven't thought about it enough

And to Aerijo

- Sending request on dirty documents is VERY surprising and as you said impossible to handle.

**Comment by Aerijo** [comment]
@LevoD Thanks for clearing that up; it's easier to understand what you want now. I didn't understand originally, because it doesn't make any sense to me for the server to be providing information about previous states of the document when the current may be arbitrarily different.

But if I understand right, you just want the spec to specify that if sync is disabled, the client should never send requests when the document is dirty? I think that's worth discussing.

> Sending request on dirty documents is VERY surprising and as you said impossible to handle

Yes, but sending requests based on previous states of the document equally surprises me. Perhaps you could share an example of where this would apply, and how the server can't track the changes and resolve them itself?

**Comment by LevoD** [comment]
> Thanks for clearing that up

This is my first time posting an issue. I'm glad I didn't botch it too bad and became more clear

> But if I understand right, you just want the spec to specify that if sync is disabled, the client should never send requests when the document is dirty? I think that's worth discussing.

Yes exactly

> Yes, but sending requests based on previous states of the document equally surprises me

It was an idea that for compatibility with servers that don't sync to send a request based on the original file. However it was just a thought. Specifying sync only to ignore request sounds pretty crazy to me which is what I need to do for the moment

> Perhaps you could share an example of where this would apply, and how the server can't track the changes and resolve them itself 

It could track some changes for the moment but right now it would just to be to ignore dirty files. In the future it'd can try to figure out if the statement exist in the original file and pull it up from the server records.

My current example is I'm writing a language server for a language that doesn't have a public beta yet. Parsing a function to build information is at the bottom of the list right now. However we want to be able to jump to definitions, hover and do other things while the file isn't edited to make reading code easier.

It was just a thought that if the client can easily figure out where the statement exist in the file on disk it might want to do that instead of not send a request. But currently the server would have to sync in order to not send incorrect hover information

**Comment by LevoD** [comment]
@dbaeumer looks like @Aerijo and I mainly agree. Should I post another issue that is worded better or is this thread fine? In short it's bizarre that a client is sending request about a document that has been changed to a server that marked textDocumentSync as none.

**Comment by dbaeumer** [comment]
No, we can reopen this issue.

However I am reluctant to add this to the spec and enforce clients to adhere to it since it is breaking for I would say all clients that currently exists. I would rather document the state of the art that if a server declares `textDocumentSync: none` it might still receive requests for changed documents.
